### PR TITLE
Add inferencer infer

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/10364

### CODE REVIEW

File: python/paddle/fluid/inferencer.py [hunk 1]
```diff
@@ -20,27 +20,54 @@
...
         self.scope = core.Scope()
         self.place = place
-        self.startup_program = framework.Program()
-        # TODO: generate the startup_program with network_func
-
-        exe = executor.Executor(place)
-        exe.run(self.startup_program, scope=self.scope)
 
-        if param_path:
+        self.exe = executor.Executor(place)
```
Comment (JiayiFeng) [inline python/paddle/fluid/inferencer.py]: The `place` could be `None`.

File: python/paddle/fluid/inferencer.py [hunk 2]
```diff
@@ -20,27 +20,54 @@
...
+            raise ValueError("network_func and param_path can not both be None")
 
-        # 2. move the default_main_program to self.program.
-
-        # 3. run the default_startup program.
-
-        # 4. load params from param_path into scope
+        self.param_path = param_path
         self.scope = core.Scope()
         self.place = place
```
Comment (JiayiFeng) [inline python/paddle/fluid/inferencer.py]: `self.place` is never used.

File: python/paddle/fluid/inferencer.py [hunk 3]
```diff
@@ -20,27 +20,54 @@
...
+        self.exe = executor.Executor(place)
+        with executor.scope_guard(self.scope):
             # load params from param_path into scope
-            io.load_persistables(exe, dirname=param_path)
+            if param_path is not None:
+                [self.inference_program, _,
+                 self.fetch_targets] = io.load_inference_model(
+                     executor=self.exe, dirname=param_path)
+            else:
+                self.inference_program = framework.Program()
```
Comment (JiayiFeng) [inline python/paddle/fluid/inferencer.py]: Where are parameter values from if `param_path` is `None`?

File: python/paddle/fluid/trainer.py [hunk 4]
```diff
@@ -68,22 +68,32 @@ def __init__(self, program_func, optimizer, param_path=None, place=None):
...
-            loss = program_func()
             if not isinstance(optimizer, opt_module.Optimizer):
                 raise TypeError(
                     "The optimizer should be an instance of Optimizer")
+            out = program_func()
+            if not isinstance(out, tuple) or len(out) != 2:
+                raise ValueError(
+                    "program_func should return tuple(loss, predict_vars)")
+
+            loss = out[0]
```
Comment (JiayiFeng) [inline python/paddle/fluid/trainer.py]: Here is a problem I also encounter in my `trainer.test()`. The `program_func()` may returns more than one element. But only one of them('loss') will be used to build the backward part. In the current design, we have no method to point out which one is the loss.

Comment (jacquesqiao) [reply python/paddle/fluid/trainer.py]: I create a issue for this problem https://github.com/PaddlePaddle/Paddle/issues/10474.

File: python/paddle/fluid/inferencer.py (hunk 1 above)
Comment (jacquesqiao) [reply python/paddle/fluid/inferencer.py]: fixed

File: python/paddle/fluid/inferencer.py (hunk 2 above)
Comment (jacquesqiao) [reply python/paddle/fluid/inferencer.py]: removed

File: python/paddle/fluid/trainer.py [hunk 5]
```diff
@@ -177,31 +209,14 @@ def save_params(self, param_path):
...
-                raise TypeError("Place should be either CUDAPlace or CPUPlace")
-            return place
+    def save_inference_model(self, model_path):
+        inference_program = framework.Program()
+        with framework.program_guard(inference_program):
+            with unique_name.guard():
+                predict_var = self.infer_func()
+        predict_var = self.train_program.block(0).var(predict_var.name)
+        exe = executor.Executor(self.place)
+        io.save_inference_model(model_path, [], [predict_var], exe)
```
Comment (JiayiFeng) [inline python/paddle/fluid/trainer.py]: Here the `predict_var` is in the `train_program`, so we are using the `train_program` to build the inference model. Why not use `inference_program` directly?

File: python/paddle/fluid/inferencer.py [hunk 6]
```diff
@@ -16,31 +16,42 @@
...
-    def __init__(self, network_func, param_path=None, place=None):
-        # 1. we need to generate a framework.Program by calling
-        # network_func. Reference: fluid.program_guard in test_word2vec.py
-
-        # 2. move the default_main_program to self.program.
-
-        # 3. run the default_startup program.
-
-        # 4. load params from param_path into scope
+    def __init__(self, param_path, place=None):
```
Comment (jetfuel) [inline python/paddle/fluid/inferencer.py]: So Inferencer doesn't need to take the Infer_program anymore?

File: python/paddle/fluid/trainer.py (hunk 5 above)
Comment (jacquesqiao) [reply python/paddle/fluid/trainer.py]: Because save_inference_model will do some other thing to inference_program, we need to rewrite it, but currently, I think rewrite this function do not influence the interface and may introduce bugs.

File: python/paddle/fluid/trainer.py [hunk 7]
```diff
@@ -56,35 +56,67 @@ def __init__(self, epoch_id, step_id):
...
+        train_func(callable): A function which will return loss. The loss must be a scalar.
+        infer_func(callable): A function which will return predict, used to save inference model
         optimizer(optimizer.Optimizer): The optimizer should be an instance of Optimizer
         place: The device place of this trainer.
     """
 
-    def __init__(self, program_func, optimizer, param_path=None, place=None):
+    def __init__(self,
+                 train_func,
+                 infer_func,
```
[the remark this answers is not in the document]
Comment (jacquesqiao) [reply python/paddle/fluid/trainer.py]: because Trainer.save_inference_model need to know the inference program. infer_func is used to get the inference program.

File: python/paddle/fluid/io.py [hunk 8]
```diff
@@ -323,9 +326,10 @@ def save_inference_model(dirname,
     if isinstance(feeded_var_names, basestring):
         feeded_var_names = [feeded_var_names]
     else:
-        if not (bool(feeded_var_names) and all(
-                isinstance(name, basestring) for name in feeded_var_names)):
-            raise ValueError("'feed_var_names' should be a list of str.")
+        if len(feeded_var_names) > 0:
```
Comment (daming-lu) [inline python/paddle/fluid/io.py]: Should we throw error if `feeded_var_names` is not an array? Also I found the correct spell is `fed_var_names`, not `feeded_`. And it is being used in many places.

File: python/paddle/fluid/tests/book/high-level-api/word2vec/no_test_word2vec_new_api.py [hunk 9]
```diff
@@ -94,44 +94,44 @@ def train(use_cuda, is_sparse, save_path):
...
                 sys.exit("got NaN loss, training failed.")
 
     trainer = fluid.Trainer(
         partial(train_program, is_sparse),
+        partial(inference_program, is_sparse),
         fluid.optimizer.SGD(learning_rate=0.001),
         place=place)
     trainer.train(
-        reader=train_reader, num_epochs=100, event_handler=event_handler)
+        reader=train_reader, num_epochs=1, event_handler=event_handler)
```
Comment (daming-lu) [inline python/paddle/fluid/tests/book/high-level-api/word2vec/no_test_word2vec_new_api.py]: So the default `num_epochs` is 1?

Comment (jacquesqiao) [reply python/paddle/fluid/tests/book/high-level-api/word2vec/no_test_word2vec_new_api.py]: this is for test, I can change it back to 100

File: python/paddle/fluid/io.py (hunk 8 above)
Comment (jacquesqiao) [reply python/paddle/fluid/io.py]: oh, feeded wrong, can you help to correct it. I think `feed_var_names` is ok.
